README.md
# plugin-placesmenu

### Simple Places Menu for LXQt-Panel similar to [Places Status Indicator(Gnome Shell)](https://extensions.gnome.org/extension/8/places-status-indicator/)

#### This project has no release versions, only the master branch.

![plugin-placesmenu](http://alexandrecvieira.droppages.com/images/indicator-places/plugin-placesmenu.png)

##### Removable media notifications

<img alt="plugin-placesmenu-connect-message" title="plugin-placesmenu Connect Message" src="http://alexandrecvieira.droppages.com/images/indicator-places/plugin-placesmenu-connect-message.png" width="320" /> <img alt="plugin-placesmenu-eject-message" title="plugin-placesmenu Eject Message" src="http://alexandrecvieira.droppages.com/images/indicator-places/plugin-placesmenu-eject-message.png" width="320" />

![plugin-placesmenu-busy-message](http://alexandrecvieira.droppages.com/images/indicator-places/plugin-placesmenu-busy-message.png)

#### Dependencies
    sudo apt install lxqt-build-tools libqt5x11extras5-dev liblxqt0-dev \
	liblxqt-globalkeys0-dev liblxqt-globalkeys-ui0-dev libkf5windowsystem-dev \
	libfm-qt-dev libglib2.0-dev libnotify-dev

#### Installation
	git clone https://github.com/alexandrecvieira/plugin-placesmenu.git
	cd plugin-placesmenu
	mkdir build
    cd build
    cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_DATAROOTDIR=/usr/share \
	-DCMAKE_INSTALL_LIBDIR=/usr/lib/x86_64-linux-gnu \
	-DCMAKE_MODULE_PATH=/usr/share/cmake/lxqt-build-tools/find-modules/
    make
    sudo make install


ChangeLog

2020-02-02  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Some volumes cannot be ejected, only unmounted

2020-01-31  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Busy notification

2020-01-28  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Mount operation interactive

2020-01-24  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Code improvements
	* Complete removable media menu

2020-01-23  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Submenus: Open and Eject

2020-01-22  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Removable media notification

2020-01-21  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* notification with libnotify

2020-01-16  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Now shows network drives too
	* Starting to include removable drives

2020-01-14  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Author

2020-01-13  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Deleted unnecessary lines

2020-01-11  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Dropped menu duplicity

2020-01-10  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Included bookmarks and fm-qt library

2019-12-15  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Dropped configuration dialog
	* To LXQt-0.14.1

2017-01-05  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Merge branch 'master' of https://github.com/acamargovieira/plugin-placesmenu
	* Retirado icone para que apareca somente o label Places

2017-01-04  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Update
	* Desenvolvimento iniciado e em fase avançada, testada a funcionalidade principal


placesmenu.h
/* BEGIN_COMMON_COPYRIGHT_HEADER
 * (c)LGPL2+
 *
 * LXDE-Qt - a lightweight, Qt based, desktop toolset
 * http://razor-qt.org
 * http://lxqt.org
 *
 * Copyright: 2015 LXQt team
 * Authors:
 *   Alexandre C Vieira <acamargo.vieira@gmail.com>
 *
 * This program or library is free software; you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 *
 * END_COMMON_COPYRIGHT_HEADER */


#ifndef PLACESMENU_H
#define PLACESMENU_H

#include "ilxqtpanelplugin.h"
#include <vector>
#include <QLabel>
#include <QToolButton>
#include <QDomElement>
#include <QAction>
#include <QDir>
#include <QSignalMapper>
#include <QSettings>
#include <QMenu>
#include <libfm-qt/core/bookmarks.h>
#include <libfm-qt/core/filepath.h>

class PlacesMenu :  public QObject, public ILXQtPanelPlugin
{
    Q_OBJECT

public:
    PlacesMenu(const ILXQtPanelPluginStartupInfo& startupInfo);
    ~PlacesMenu();

    virtual QWidget* widget() { return &mButton; }
    virtual QString themeId() const { return "PlacesMenu"; }
    virtual Flags flags() const { return SingleInstance; }
    
private Q_SLOTS:
    void showMenu();
    void openDirectory(const QString& path);
    void createMenuItem(QMenu* menu, const QString& name, const QString& iconName, const QString& location);
    void createMenuItem(QMenu* menu, const QString& name, QIcon icon,  const QString& location);
    void createMenuEject(QMenu* menu, const QString& name, const QString& volmeName);
    void createSubmenu(QMenu* menu, GMount* mount);
    void createSubmenu(QMenu* menu, GVolume* volume);
    void onEject(const QString& volumeName);

protected Q_SLOTS:
    void buildMenu();

protected:
    static void onVolumeAdded(GVolumeMonitor* monitor, GVolume* volume, PlacesMenu* pThis);
    static void onVolumeRemoved(GVolumeMonitor* monitor, GVolume* volume, PlacesMenu* pThis);
    static void showMessage(const QString& text);
    static void onEjectVolumeFinished(GVolume* volume, GAsyncResult* res, PlacesMenu* /*pThis*/);

private:
    void addActions(QMenu* menu);
       
    QToolButton mButton;
    QMenu* mMenu;
    QSignalMapper* mOpenDirectorySignalMapper;
    QSignalMapper* mEjectSignalMapper;

    QDir mBaseDirectory;
    const QString mDefaultIcon = "folder";
    std::vector<QString> mPathStrings;
    std::shared_ptr<Fm::Bookmarks> bookmarks_;
    QMap<QString, GVolume*> mapVolumes;

    GVolumeMonitor* volumeMonitor;
};

class PlacesMenuLibrary: public QObject, public ILXQtPanelPluginLibrary
{
    Q_OBJECT
    Q_PLUGIN_METADATA(IID "lxde-qt.org/Panel/PluginInterface/3.0")
    Q_INTERFACES(ILXQtPanelPluginLibrary)
public:
    ILXQtPanelPlugin* instance(const ILXQtPanelPluginStartupInfo &startupInfo) const
    {
        return new PlacesMenu(startupInfo);
    }
};


#endif



placesmenuconfiguration.h
/* BEGIN_COMMON_COPYRIGHT_HEADER
 * (c)LGPL2+
 *
 * LXDE-Qt - a lightweight, Qt based, desktop toolset
 * http://razor-qt.org
 * http://lxqt.org
 *
 * Copyright: 2015 LXQt team
 * Authors:
 *   Daniel Drzisga <sersmicro@gmail.com>
 *
 * This program or library is free software; you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 *
 * END_COMMON_COPYRIGHT_HEADER */


#ifndef PLACESMENUCONFIGURATION_H
#define PLACESMENUCONFIGURATION_H

#include "lxqtpanelpluginconfigdialog.h"
#include "pluginsettings.h"

#include <QAbstractButton>
#include <QButtonGroup>
#include <QLocale>
#include <QDateTime>
#include <QDir>

namespace Ui {
    class PlacesMenuConfiguration;
}

class PlacesMenuConfiguration : public LXQtPanelPluginConfigDialog
{
    Q_OBJECT

public:
    explicit PlacesMenuConfiguration(PluginSettings *settings, QWidget *parent = nullptr);
    ~PlacesMenuConfiguration();

private:
    Ui::PlacesMenuConfiguration *ui;
    QDir mBaseDirectory;
    QString mIcon;
    QIcon mDefaultIcon;

    /*
      Read settings from conf file and put data into controls.
    */
   void loadSettings();

private slots:
    /*
      Saves settings in conf file.
    */
    void saveSettings();
    void showDirectoryDialog();
    void showIconDialog();

private:
};

#endif // PLACESMENUCONFIGURATION_H



translations/placesmenu.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="en_US">
<context>
    <name>PlacesMenu</name>
    <message>
        <location filename="../placesmenu.cpp" line="54"/>
        <source>Places</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="116"/>
        <source>Open</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="117"/>
        <source>Eject removable media</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="155"/>
        <source>Home</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="159"/>
        <source>Documents</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="167"/>
        <source>Music</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="171"/>
        <source>Pictures</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="175"/>
        <source>Videos</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="163"/>
        <source>Downloads</source>
        <translation type="unfinished"></translation>
    </message>
     <message>
        <location filename="../placesmenu.cpp" line="213"/>
        <source>The device &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; is connected.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="222"/>
        <source>The device &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; is removed.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="230"/>
        <source>Places Menu</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="256"/>
        <source>The device &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; is busy, can not be removed.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>


placesmenuconfiguration.cpp
/* BEGIN_COMMON_COPYRIGHT_HEADER
 * (c)LGPL2+
 *
 * LXDE-Qt - a lightweight, Qt based, desktop toolset
 * http://razor-qt.org
 * http://lxqt.org
 *
 * Copyright: 2015 LXQt team
 * Authors:
 *   Daniel Drzisga <sersmicro@gmail.com>
 *
 * This program or library is free software; you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 *
 * END_COMMON_COPYRIGHT_HEADER */


#include <QIcon>
#include <QInputDialog>
#include <QFileDialog>
#include <QMessageBox>

#include <XdgIcon>

#include "placesmenuconfiguration.h"
#include "ui_placesmenuconfiguration.h"


PlacesMenuConfiguration::PlacesMenuConfiguration(PluginSettings *settings, QWidget *parent) :
    LXQtPanelPluginConfigDialog(settings, parent),
    ui(new Ui::PlacesMenuConfiguration),
    mBaseDirectory(QDir::homePath()),
    mDefaultIcon(XdgIcon::fromTheme("folder"))
{
    setAttribute(Qt::WA_DeleteOnClose);
    setObjectName("PlacesMenuConfigurationWindow");
    ui->setupUi(this);

    connect(ui->buttons, SIGNAL(clicked(QAbstractButton*)), SLOT(dialogButtonsAction(QAbstractButton*)));

    loadSettings();
    ui->baseDirectoryB->setIcon(mDefaultIcon);

    connect(ui->baseDirectoryB, SIGNAL(clicked()), SLOT(showDirectoryDialog()));
    connect(ui->iconB, SIGNAL(clicked()), SLOT(showIconDialog()));
}

PlacesMenuConfiguration::~PlacesMenuConfiguration()
{
    delete ui;
}

void PlacesMenuConfiguration::loadSettings()
{
    mBaseDirectory.setPath(settings().value("baseDirectory", QDir::homePath()).toString());
    ui->baseDirectoryB->setText(mBaseDirectory.dirName());

    mIcon = settings().value("icon", QString()).toString();
    if(!mIcon.isNull())
    {
        QIcon buttonIcon = QIcon(mIcon);
        if(!buttonIcon.pixmap(QSize(24,24)).isNull())
        {
            ui->iconB->setIcon(buttonIcon);
            return;
        }
    }

    ui->iconB->setIcon(mDefaultIcon);
}

void PlacesMenuConfiguration::saveSettings()
{
    settings().setValue("baseDirectory", mBaseDirectory.absolutePath());
    settings().setValue("icon", mIcon);
}

void PlacesMenuConfiguration::showDirectoryDialog()
{
    QFileDialog d(this, tr("Choose Base Directory"), mBaseDirectory.absolutePath());
    d.setFileMode(QFileDialog::Directory);
    d.setOptions(QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
    d.setWindowModality(Qt::WindowModal);

    if(d.exec() && !d.selectedFiles().isEmpty())
    {
        mBaseDirectory.setPath(d.selectedFiles().front());
        ui->baseDirectoryB->setText(mBaseDirectory.dirName());

        saveSettings();
    }
}

void PlacesMenuConfiguration::showIconDialog()
{
    QFileDialog d(this, tr("Choose Icon"), QDir::homePath(), tr("Icons (*.png *.xpm *.jpg)"));
    d.setWindowModality(Qt::WindowModal);

    if(d.exec() && !d.selectedFiles().isEmpty())
    {
        QIcon newIcon = QIcon(d.selectedFiles().front());

        if(newIcon.pixmap(QSize(24,24)).isNull())
        {
            QMessageBox::warning(this, tr("Directory Menu"), tr("An error occurred while loading the icon."));
            return;
        }

        ui->iconB->setIcon(newIcon);
        mIcon = d.selectedFiles().front();
        saveSettings();
    }
}



translations/placesmenu_pt_BR.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="pt_BR">
<context>
    <name>PlacesMenu</name>
    <message>
        <location filename="../placesmenu.cpp" line="54"/>
        <source>Places</source>
        <translation>Locais</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="116"/>
        <source>Open</source>
        <translation>Abrir</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="117"/>
        <source>Eject removable media</source>
        <translation>Ejetar mídia removível</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="155"/>
        <source>Home</source>
        <translation>Pasta pessoal</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="159"/>
        <source>Documents</source>
        <translation>Documentos</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="167"/>
        <source>Music</source>
        <translation>Música</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="171"/>
        <source>Pictures</source>
        <translation>Imagens</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="175"/>
        <source>Videos</source>
        <translation>Vídeos</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="163"/>
        <source>Downloads</source>
        <translation>Downloads</translation>
    </message>
     <message>
        <location filename="../placesmenu.cpp" line="213"/>
        <source>The device &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; is connected.</source>
        <translation>O dispositivo &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; foi conectado.</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="222"/>
        <source>The device &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; is removed.</source>
        <translation>O dispositivo &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; foi removido.</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="230"/>
        <source>Places Menu</source>
        <translation>Menu Locais</translation>
    </message>
    <message>
        <location filename="../placesmenu.cpp" line="256"/>
        <source>The device &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; is busy, can not be removed.</source>
        <translation>O dispositivo &lt;b&gt;&lt;nobr&gt;&quot;%1&quot;&lt;/nobr&gt;&lt;/b&gt; está ocupado, não pode ser removido.</translation>
    </message>
</context>
</TS>


placesmenu.cpp
/* BEGIN_COMMON_COPYRIGHT_HEADER
 * (c)LGPL2+
 *
 * LXDE-Qt - a lightweight, Qt based, desktop toolset
 * http://razor-qt.org
 * http://lxqt.org
 *
 * Copyright: 2019
 * Authors:
 *   Alexandre C Vieira <acamargo.vieira@gmail.com>
 *
 * This program or library is free software; you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is diinstributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 *
 * END_COMMON_COPYRIGHT_HEADER */

#include "placesmenu.h"
#include <QStandardPaths>
#include <QStyle>
#include <QDebug>
#include <QDesktopServices>
#include <QFileInfo>
#include <QUrl>
#include <QIcon>
#include <XdgIcon>
#include <libnotify/notify.h>

PlacesMenu::PlacesMenu(const ILXQtPanelPluginStartupInfo &startupInfo) :
    QObject(),
    ILXQtPanelPlugin(startupInfo),
    mMenu(0),
    bookmarks_{Fm::Bookmarks::globalInstance()}
{
    // ensure that glib integration of Qt is not turned off
    // This fixes #168: https://github.com/lxqt/pcmanfm-qt/issues/168
    qunsetenv("QT_NO_GLIB");

    mOpenDirectorySignalMapper = new QSignalMapper(this);
    mEjectSignalMapper = new QSignalMapper(this);

    mButton.setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Minimum);
    mButton.setText(QString(tr("Places")));
    
    connect(&mButton, SIGNAL(clicked()), this, SLOT(showMenu()));
    connect(mOpenDirectorySignalMapper, SIGNAL(mapped(QString)), this, SLOT(openDirectory(QString)));
    connect(mEjectSignalMapper, SIGNAL(mapped(QString)), this, SLOT(onEject(QString)));
    
    volumeMonitor = g_volume_monitor_get();
    g_signal_connect(volumeMonitor, "volume-added", G_CALLBACK(onVolumeAdded), this);
    g_signal_connect(volumeMonitor, "volume-removed", G_CALLBACK(onVolumeRemoved), this);
}

PlacesMenu::~PlacesMenu()
{
    if(volumeMonitor) {
        g_signal_handlers_disconnect_by_func(volumeMonitor, (gpointer)G_CALLBACK(onVolumeAdded), this);
        g_signal_handlers_disconnect_by_func(volumeMonitor, (gpointer)G_CALLBACK(onVolumeRemoved), this);
    }
    
    if(mMenu)
    {
        delete mMenu;
        mMenu = 0;
    }
}

void PlacesMenu::showMenu()
{
    buildMenu();
    
    willShowWindow(mMenu);
    // Just using Qt`s activateWindow() won't work on some WMs like Kwin.
    // Solution is to execute menu 1ms later using timer
    mMenu->popup(calculatePopupWindowPos(mMenu->sizeHint()).topLeft());
}

void PlacesMenu::buildMenu()
{
    if(mMenu)
    {
        delete mMenu;
        mMenu = 0;
    }

    mMenu = new QMenu();

    addActions(mMenu);
}

void PlacesMenu::openDirectory(const QString& path)
{
    QDesktopServices::openUrl(QUrl("file://" + QDir::toNativeSeparators(path)));
}

void PlacesMenu::createSubmenu(QMenu* menu, GVolume* volume)
{
    QString volumeName = QString::fromUtf8(g_volume_get_name(volume));
    GIcon* gicon = g_volume_get_icon(volume);
    QIcon icon = Fm::IconInfo::fromGIcon(gicon)->qicon();
    QMenu* subMenu = menu->addMenu(icon, volumeName);
    GFile* mountFile = g_mount_get_default_location(g_volume_get_mount(volume));
    QString mountPath = QString::fromUtf8(g_file_get_path(mountFile));
    createMenuItem(subMenu, tr("Open"), mDefaultIcon, mountPath);
    createMenuEject(subMenu, tr("Eject removable media"), volumeName);
    mapVolumes.insert(volumeName, volume);
}

void PlacesMenu::createSubmenu(QMenu* menu, GMount* mount)
{
    QString mountName = QString::fromUtf8(g_mount_get_name(mount));
    GIcon* gicon = g_mount_get_icon(mount);
    QIcon icon = Fm::IconInfo::fromGIcon(gicon)->qicon();
    GFile* mountFile = g_mount_get_default_location(mount);
    QString mountPath = QString::fromUtf8(g_file_get_path(mountFile));
    createMenuItem(menu, mountName, icon, mountPath);
}

void PlacesMenu::createMenuItem(QMenu* menu, const QString& name, const QString& iconName,  const QString& location)
{
    QAction* action = menu->addAction(XdgIcon::fromTheme(iconName), name);
    connect(action, SIGNAL(triggered()), mOpenDirectorySignalMapper, SLOT(map()));
    mOpenDirectorySignalMapper->setMapping(action, location);    
}

void PlacesMenu::createMenuItem(QMenu* menu, const QString& name, QIcon icon,  const QString& location)
{
    QAction* action = menu->addAction(icon, name);
    connect(action, SIGNAL(triggered()), mOpenDirectorySignalMapper, SLOT(map()));
    mOpenDirectorySignalMapper->setMapping(action, location);    
}

void PlacesMenu::createMenuEject(QMenu* menu,  const QString& name, const QString& volumeName)
{
    QAction* action = menu->addAction(XdgIcon::fromTheme("media-eject"), name);
    connect(action, SIGNAL(triggered()), mEjectSignalMapper, SLOT(map()));
    mEjectSignalMapper->setMapping(action, volumeName);
}

void PlacesMenu::addActions(QMenu* menu)
{
    QString homeName = tr("Home");
    QString homeLocation = QStandardPaths::locate(QStandardPaths::HomeLocation, QString(), QStandardPaths::LocateDirectory);
    createMenuItem(menu, homeName, "user-home", homeLocation);
       
    QString documentsName = tr("Documents");
    QString documentsLocation = QStandardPaths::locate(QStandardPaths::DocumentsLocation, QString(), QStandardPaths::LocateDirectory);
    createMenuItem(menu, documentsName, "folder-documents", documentsLocation);
     
    QString downloadName = tr("Downloads");
    QString downloadLocation = QStandardPaths::locate(QStandardPaths::DownloadLocation, QString(), QStandardPaths::LocateDirectory);
    createMenuItem(menu, downloadName, "folder-download", downloadLocation);
    
    QString musicName = tr("Music");
    QString musicLocation = QStandardPaths::locate(QStandardPaths::MusicLocation, QString(), QStandardPaths::LocateDirectory);
    createMenuItem(menu, musicName, "folder-music", musicLocation);
    
    QString pictureName = tr("Pictures");
    QString pictureLocation = QStandardPaths::locate(QStandardPaths::PicturesLocation, QString(), QStandardPaths::LocateDirectory);
    createMenuItem(menu, pictureName, "folder-picture", pictureLocation);
    
    QString videosName = tr("Videos");
    QString videosLocation = QStandardPaths::locate(QStandardPaths::MoviesLocation, QString(), QStandardPaths::LocateDirectory);
    createMenuItem(menu, videosName, "folder-videos", videosLocation);
    
    menu->addSeparator();

    // load bookmarks
    for(auto& item: bookmarks_->items()) {
     	QString bookmarkName = item->name();
	auto bookmarkPath = item->path().localPath();
	QString bookmarkLocation = QString::fromUtf8(bookmarkPath.get());
	bookmarkLocation += "/";
	if((bookmarkLocation != documentsLocation) && (bookmarkLocation != downloadLocation)
	   && (bookmarkLocation != musicLocation) && (bookmarkLocation != pictureLocation)
	   && (bookmarkLocation != videosLocation)){
	    createMenuItem(menu, bookmarkName, mDefaultIcon, bookmarkLocation);
	}
    }

    // Mounted drives: populate mounted drives and connect
    int count = 0;
    GList* mountsList = g_volume_monitor_get_mounts(volumeMonitor);
    for(GList* l = mountsList; l; l = l->next) {
	GMount* mount = G_MOUNT(l->data);
	GVolume* volume = g_mount_get_volume(mount);
	count++;
	if(count == 1)
	    menu->addSeparator();
	if(G_IS_VOLUME(volume) && g_volume_can_eject(volume))
	    createSubmenu(menu, volume);
	else
	    createSubmenu(menu, mount);
    }
}

void PlacesMenu::onVolumeAdded(GVolumeMonitor* /*monitor*/, GVolume* volume, PlacesMenu* /*pThis*/)
{
    char* volumeName = g_volume_get_name(volume);
    QString text = tr("The device <b><nobr>\"%1\"</nobr></b> is connected.").arg(volumeName);
    showMessage(text);
   
    g_free(volumeName);
}

void PlacesMenu::onVolumeRemoved(GVolumeMonitor* /*monitor*/, GVolume* volume, PlacesMenu* /*pThis*/)
{
    char* volumeName = g_volume_get_name(volume);
    QString text = tr("The device <b><nobr>\"%1\"</nobr></b> is removed.").arg(volumeName);
    showMessage(text);
    
    g_free(volumeName);
}

void PlacesMenu::showMessage(const QString& text)
{
    QString title = tr("Places Menu");
    notify_init(title.toUtf8().data());
    NotifyNotification* notification = notify_notification_new(text.toUtf8().data(), nullptr, nullptr);
    notify_notification_set_urgency(notification, NOTIFY_URGENCY_NORMAL);
    notify_notification_set_timeout(notification, NOTIFY_EXPIRES_DEFAULT);
    notify_notification_show(notification, NULL);

    g_object_unref(G_OBJECT(notification));

    notify_uninit();
}

void PlacesMenu::onEject(const QString& volumeName)
{
    GVolume* volume = mapVolumes[volumeName];
    mapVolumes.remove(volumeName);
    g_volume_eject_with_operation(volume, G_MOUNT_UNMOUNT_NONE, nullptr, nullptr, (GAsyncReadyCallback)onEjectVolumeFinished, nullptr);
}

void PlacesMenu::onEjectVolumeFinished(GVolume* volume, GAsyncResult* res, PlacesMenu* /*pThis*/) {
    QString volumeName = g_volume_get_name(volume);
    
    GError* error = nullptr;
    g_volume_eject_with_operation_finish(volume, res, &error);
    
    if(error){
	QString text = tr("The device <b><nobr>\"%1\"</nobr></b> is busy, can not be removed.").arg(volumeName);
	showMessage(text);
    }
    
    if(error) {
	g_error_free(error);
    }
}

